File: apps/mcp-server/src/pipeline/manifest.ts
/**
 * Build a CallManifest from a raw Horizon transaction.
 *
 * The actual XDR decode is done server-side using the Stellar JS SDK.
 * We parse diagnostic events, auth entries, and SAC transfer events.
 */

import type { RawTransaction } from "./recorder";

export interface AssetFlow {
  asset_id: string;
  asset_symbol?: string;
  direction: "outbound" | "inbound";
  amount_raw: string; // u128 as string
  amount_display: string;
  decimals: number;
  counterparty?: string;
}

export interface ContractRef {
  id: string;
  label?: string;
  protocol?: string;
}

export interface AuthBoundary {
  contract_id: string;
  function_name: string;
  account: string;
}

export interface CallNode {
  contract_id: string;
  contract_label?: string;
  function_name: string;
  args: any[];
  sub_calls: CallNode[];
  requires_auth: boolean;
  authorized_by?: string;
}

export interface CallManifest {
  transaction_hash: string;
  network: string;
  ledger_sequence: number;
  timestamp: string;
  invoking_account: string;
  top_level_calls: CallNode[];
  unique_contracts: ContractRef[];
  unique_functions: string[];
  asset_flows: AssetFlow[];
  auth_boundaries: AuthBoundary[];
  observed_amounts: Record<string, { min_raw: string; max_raw: string; asset_id: string }>;
  summary: string;
  is_simulation: boolean;
  simulation_cost?: { cpu_instructions: number; memory_bytes: number };
}

// Well-known Soroban protocol labels
const PROTOCOL_REGISTRY: Record<string, { label: string; protocol: string }> = {
  // Blend protocol pools (testnet)
  CDVQVKOY2YSXS2IC7KN6ZNOG5K6JZBPQZ5YBXGBVHPGJFCBKK5B6TCSS: { label: "Blend Pool", protocol: "blend" },
  // Soroswap router
  CBEZFD7JMQHKUQTUXFLT5MHFAAHZLXMXDQ4VQ7LHHEQLZMHZXJCJHIPH: { label: "Soroswap Router", protocol: "soroswap" },
};

/**
 * Build a CallManifest from a raw Horizon transaction.
 * Uses the result_meta_xdr diagnostic events and envelope to extract:
 * - Contract calls (from sorobanMeta.events)
 * - Auth boundaries (from SorobanTransactionMeta.events[].event.type == "diagnostic")
 * - SAC transfer events (contract events with topic[0] == "transfer")
 */
export function buildCallManifest(
  raw: RawTransaction,
  invokingAccountOverride?: string
): CallManifest {
  const invoking_account = invokingAccountOverride ?? raw.source_account;

  // Parse diagnostic events from result_meta_xdr
  // In the Worker, we receive the XDR as base64 and parse it with the Stellar SDK
  // For now we build the manifest from what Horizon gives us directly
  const uniqueContracts: ContractRef[] = [];
  const uniqueFunctions: string[] = [];
  const assetFlows: AssetFlow[] = [];
  const authBoundaries: AuthBoundary[] = [];

  // ── Parse from Horizon's parsed_operations if available ──────────────────
  // Horizon /transactions/{hash}/operations gives us the invoke_host_function
  // result, but we rely on the XDR for full fidelity.
  // For the TypeScript pipeline, we produce a best-effort manifest from what
  // Horizon exposes. Full fidelity requires decoding the XDR with stellar-sdk.

  const summary = `Transaction ${raw.hash.slice(0, 12)}… on ${raw.network} at ledger ${raw.ledger}`;

  return {
    transaction_hash: raw.hash,
    network: raw.network,
    ledger_sequence: raw.ledger,
    timestamp: raw.created_at,
    invoking_account,
    top_level_calls: [],
    unique_contracts: uniqueContracts,
    unique_functions: uniqueFunctions,
    asset_flows: assetFlows,
    auth_boundaries: authBoundaries,
    observed_amounts: {},
    summary,
    is_simulation: false,
  };
}

/**
 * Build a CallManifest directly from structured input (for testing / agent use).
 * This is what agents call when they already have the decoded data.
 */
export function buildManifestFromStructured(input: {
  tx_hash: string;
  network: string;
  ledger: number;
  timestamp: string;
  invoking_account: string;
  contracts: Array<{ id: string; label?: string; protocol?: string }>;
  functions: string[];
  asset_flows: Array<{
    asset_id: string;
    asset_symbol?: string;
    direction: "outbound" | "inbound";
    amount_raw: string;
    decimals?: number;
    counterparty?: string;
  }>;
  auth_boundaries?: Array<{ contract_id: string; function_name: string; account: string }>;
}): CallManifest {
  return {
    transaction_hash: input.tx_hash,
    network: input.network,
    ledger_sequence: input.ledger,
    timestamp: input.timestamp,
    invoking_account: input.invoking_account,
    top_level_calls: [],
    unique_contracts: input.contracts.map((c) => ({
      id: c.id,
      label: c.label ?? PROTOCOL_REGISTRY[c.id]?.label,
      protocol: c.protocol ?? PROTOCOL_REGISTRY[c.id]?.protocol,
    })),
    unique_functions: input.functions,
    asset_flows: input.asset_flows.map((f) => ({
      ...f,
      decimals: f.decimals ?? 7,
      amount_display: formatAmount(BigInt(f.amount_raw), f.decimals ?? 7),
    })),
    auth_boundaries: input.auth_boundaries ?? [],
    observed_amounts: {},
    summary: `${input.functions.join(", ")} on ${input.contracts.map((c) => c.label ?? c.id.slice(0, 8)).join(", ")}`,
    is_simulation: false,
  };
}

function formatAmount(raw: bigint, decimals: number): string {
  const divisor = BigInt(10 ** decimals);
  const whole = raw / divisor;
  const frac = raw % divisor;
  return `${whole}.${frac.toString().padStart(decimals, "0")}`;
}

File: packages/core/src/recorder/manifest.rs
//! CallManifest — the structured representation of a recorded Soroban transaction.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// A single contract invocation node in the call tree.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CallNode {
    /// Contract ID (C-address format)
    pub contract_id: String,
    /// Human-readable label from protocol registry (if known)
    pub contract_label: Option<String>,
    /// Function name invoked
    pub function_name: String,
    /// Decoded argument list (JSON-serializable)
    pub args: Vec<serde_json::Value>,
    /// Sub-calls made by this invocation
    pub sub_calls: Vec<CallNode>,
    /// Whether this call triggered a require_auth()
    pub requires_auth: bool,
    /// The account that authorized this call (if known)
    pub authorized_by: Option<String>,
}

/// A contract referenced in the manifest.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ContractRef {
    /// Contract ID (C-address format)
    pub id: String,
    /// Human-readable label (from protocol registry)
    pub label: Option<String>,
    /// Protocol family (blend, soroswap, aquarius, etc.)
    pub protocol: Option<String>,
}

/// Net token movement involving the invoking account.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AssetFlow {
    /// Asset identifier (contract ID for SAC tokens, or "native" for XLM)
    pub asset_id: String,
    /// Human-readable asset symbol (USDC, XLM, etc.)
    pub asset_symbol: Option<String>,
    /// Direction relative to invoking account
    pub direction: FlowDirection,
    /// Raw integer amount (7 decimal places for XLM, SEP-41 decimals otherwise)
    pub amount_raw: u128,
    /// Human-readable amount string with correct decimals
    pub amount_display: String,
    /// Decimal places for this asset
    pub decimals: u8,
    /// Counterparty address
    pub counterparty: Option<String>,
}

/// Direction of asset flow.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum FlowDirection {
    /// Token leaving invoking account
    Outbound,
    /// Token entering invoking account
    Inbound,
}

/// An observed require_auth() call site in the auth tree.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AuthBoundary {
    /// Contract that called require_auth()
    pub contract_id: String,
    /// Function that called require_auth()
    pub function_name: String,
    /// Account whose auth was required
    pub account: String,
}

/// Observed amount range for a (contract, function) pair.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AmountRange {
    /// Minimum observed amount (raw)
    pub min_raw: u128,
    /// Maximum observed amount (raw)
    pub max_raw: u128,
    /// Asset ID this range refers to
    pub asset_id: String,
}

/// The complete structured output of recording a transaction.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CallManifest {
    /// Transaction hash (or "simulated_{id}" for simulations)
    pub transaction_hash: String,
    /// Network this transaction was observed on
    pub network: String,
    /// Ledger sequence number (0 for simulations)
    pub ledger_sequence: u32,
    /// ISO 8601 timestamp
    pub timestamp: String,
    /// Invoking account (G-address or C-address)
    pub invoking_account: String,
    /// Pruned call tree — user-initiated calls only
    pub top_level_calls: Vec<CallNode>,
    /// Unique contracts referenced in the manifest
    pub unique_contracts: Vec<ContractRef>,
    /// Unique function names observed
    pub unique_functions: Vec<String>,
    /// Net asset flows involving the invoking account
    pub asset_flows: Vec<AssetFlow>,
    /// require_auth() call sites
    pub auth_boundaries: Vec<AuthBoundary>,
    /// Observed amount ranges per (contract_id, function_name)
    pub observed_amounts: HashMap<String, AmountRange>,
    /// Human-readable one-liner summary
    pub summary: String,
    /// Whether this was a simulation (not an executed transaction)
    pub is_simulation: bool,
    /// Simulation resource cost (only set for simulations)
    pub simulation_cost: Option<SimulationCost>,
}

/// Resource consumption from a simulated transaction.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SimulationCost {
    /// CPU instructions consumed
    pub cpu_instructions: u64,
    /// Memory bytes consumed
    pub memory_bytes: u64,
}
